segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -1040,25 +1039,29 @@ public void readTableEntriesDelta(WireCommands.ReadTableEntriesDelta readTableEn
 
     }
 
-    private void truncateIndexSegment(String segment, long offset) {
+    private CompletableFuture<Void> truncateIndexSegment(String segment, long offset) {
         String indexSegment = getIndexSegmentName(segment);
+        long indexSegmentOffset;
         try {
-            long indexSegmentOffset = IndexRequestProcessor.locateOffsetForIndexSegment(segmentStore, segment, offset, false);
-            log.debug("Truncating index segment {} at offset {}.", indexSegment, indexSegmentOffset);
-            if (indexSegmentOffset == 0) {
-                return;
-            }
-            segmentStore.truncateStreamSegment(indexSegment, indexSegmentOffset, TIMEOUT)
-                    .whenComplete((v, exp) -> {
-                        if (exp != null) {
-                            log.warn("Failed to truncate index segment {} at offset {} due to", indexSegment, indexSegmentOffset, exp);
-                        } else {
-                            log.info("Successfully truncated index segment {} at offset {}", indexSegment, indexSegmentOffset);
-                        }
-                    });
+            indexSegmentOffset = IndexRequestProcessor.locateOffsetForIndexSegment(segmentStore, segment, offset, false);
+            log.info("Truncating index segment {} at offset {}.", indexSegment, indexSegmentOffset);
         } catch (Exception e) {
-            log.warn("Unable to locate offset for index segment {}  for offset {} due to ", indexSegment, offset, e);
+            Throwable ex = Exceptions.unwrap(e);
+            if (ex instanceof IllegalArgumentException || ex instanceof IllegalStateException || ex instanceof IndexRequestProcessor.SegmentTruncatedException) {
```

Comment:
I am not clear on the difference between this if path and the if and the final else path. Is there a reason this can't be unified so we only have a single if for `StreamSegmentNotExistsException`?

Reply:
In the if block, if the I am checking if the exception of type IllegalArgumentException , IllegalStateException  and IndexRequestProcessor.SegmentTruncatedException then returning failed future with the exception **CompletableFuture.failedFuture(ex);**

But in else if part I am checking of exception type of StreamSegmentNotExistsException  then returning future has completed successfully.
so, here both the case are different.

Reply:
It doesn't return `CompletableFuture.completedFuture(null);` it is doing a `throw new CompletionException(ex);` on 1058. I don't see a semantic difference between that or `return CompletableFuture.failedFuture(ex);` unless there is some need to distinguish the call throwing vs the future throwing, but looking at the caller it appears both of these go down the exact same control path to be handled.

Reply:
Addressed.